forge/db/models/Team.js

```diff
@@ -354,6 +354,87 @@ module.exports = {
                  */
                 checkInstanceStartAllowed: async function (instance) {
                     return true
+                },
+
+                /**
+                 * Checks whether the team type can be modified to the requested type.
+                 * The team must be within any limits of the target type.
+                 */
+                checkTeamTypeUpdateAllowed: async function (teamType) {
+                    // console.log(teamType.properties)
+                    await this.ensureTeamTypeExists()
+
+                    // Check the following limits:
+                    // - User count
+                    // - Device count
+                    // - Instance Type counts
+                    const errors = []
+
+                    const currentMemberCount = await this.memberCount()
+                    const targetMemberLimit = teamType.getProperty('users.limit', -1)
+                    if (targetMemberLimit !== -1 && targetMemberLimit < currentMemberCount) {
+                        errors.push({
+                            code: 'member_limit_reached',
+                            error: 'Member limit reached',
+                            limit: targetMemberLimit,
+                            count: currentMemberCount
+                        })
+                    }
+
+                    const currentDeviceCount = await this.deviceCount()
+                    const targetDeviceLimit = teamType.getProperty('devices.limit', -1)
+                    if (targetDeviceLimit !== -1 && targetDeviceLimit < currentDeviceCount) {
+                        errors.push({
+                            code: 'device_limit_reached',
+                            error: 'Device limit reached',
+                            limit: targetDeviceLimit,
+                            count: currentDeviceCount
+                        })
+                    }
+
+                    const currentInstanceCountsByType = await this.instanceCountByType()
+                    // console.log(currentInstanceCountsByType)
+                    const targetInstanceLimits = {}
+                    for (const instanceType of Object.keys(currentInstanceCountsByType)) {
+                        // console.log('checking', instanceType, teamType.properties.instances[instanceType])
```

Comment:
✂️  Lots of this over the file, I won't comment on each one

Reply:
All tidied up